### PR TITLE
types: fix HtmlElement predicate

### LINKED ISSUE #309: HtmlElement is unrecognized
I believe I have found a bug in how `$.HtmlElement ` is defined.

```html
<section id="HtmlElement"></section>
```

```js
import $ from 'sanctuary-def'

const HtmlElement = document.querySelector ('#HtmlElement')

const testCase = $.test ([])
                        ($.HtmlElement)
                        (HtmlElement)

const sanityCheck = $.test ([])
                           ($.Boolean)
                           (true)

/*
  The Console should display true since a <section> is an HTMLElement.
  All test-cases should be true
 */

console.debug ('testCase', testCase)       // false
console.debug ('sanityCheck', sanityCheck) // true
console.debug ('check', /^\[object HTML.+Element\]$/.test (Object.toString.call (HtmlElement)))
```

Live test-case: https://playcode.io/849895/
Offending line: https://github.com/sanctuary-js/sanctuary-def/blob/master/index.js#L719

This means that currently one can not use an HtmlElement as an argument.

_please ignore the silliness of this example - I realize that `HtmlElement` is not a `StrMap`_

```js
const textContent = S.flip (S.insert ('textContent'))

textContent (HtmlElement) ('text content to inserted')
```

```
Uncaught TypeError: Unrecognized value

flip :: Functor f => f (a -> b) -> a -> f b
                                   ^
                                   1

1)  <object HTMLElement> :: (no types)

The value at position 1 is not a member of any type in the environment.

The environment contains the following types:

  - Function
  - Arguments
  - Array c
  - Array2 c d
  - Boolean
  - Buffer
  - Date
  - Descending c
  - Either c d
  - Error
  - (c -> d)
  - HtmlElement
  - Identity c
  - JsMap c d
  - JsSet c
  - Maybe c
  - Module
  - Null
  - Number
  - Object
  - Pair c d
  - RegExp
  - StrMap c
  - String
  - Symbol
  - Type
  - TypeClass
  - Undefined
  - Future c d
  - ConcurrentFuture c d
  - Event
    invalidValue index.js:2666
    satisfactoryTypes index.js:1351
    assertRight index.js:2731
    wrapNext index.js:2823
```

### PR DESCRIPTION
**Pull request by davidchambers**
Fixes #309

```javascript
> Object.prototype.toString.call (document.createElement ('div'))
'[object HTMLDivElement]'

> Object.prototype.toString.call (document.createElement ('section'))
'[object HTMLElement]'
```

Before:

```javascript
> /^\[object HTML.+Element\]$/.test (Object.prototype.toString.call (document.createElement ('section')))
false
```

After:

```javascript
> /^\[object HTML.*Element\]$/.test (Object.prototype.toString.call (document.createElement ('section')))
true
```

/cc @dotnetCarpenter

### CODE REVIEW

**Comment by dotnetCarpenter** [comment]
@davidchambers 👍 

I've tested your change in Firefox and it works! But I am puzzled to how I got ... aha I see. I used `Object.toString;` and **not** `Object.prototype.toString;`.

You're fix is great :)

File: index.js [hunk 1]
```diff
@@ -716,7 +716,7 @@
     ('HtmlElement')
     ([])
     (function(x) {
-       return /^\[object HTML.+Element\]$/.test (toString.call (x));
+       return /^\[object HTML.*Element\]$/.test (toString.call (x));
```
Comment (dotnetCarpenter) [inline index.js:719]: How about disallow space here? `^\[object HTML[^\s]*Element\]$`

**Review by Avaq — APPROVED** [review]

File: index.js (hunk 1 above)
Comment (dotnetCarpenter) [reply index.js:719]: @Avaq What do you think about changing the regex to `^\[object HTML[^\s]*Element\]$` in order to disallow spaces after `HTML`?

So `HTML ElementDiv` would fail while `HTML.ElementDiv` would be allowed.

Comment (davidchambers) [reply index.js:719]: We will accept infinitely many invalid strings in either case; I'm inclined to stick with the simpler `.*`.

By the way, `\S` is shorthand for `[^\s]`. :)

Comment (Avaq) [reply index.js:719]: > We will accept infinitely many invalid strings in either case

That was my thinking too.

Comment (dotnetCarpenter) [reply index.js:719]: I guess, since `HTMLElement` is an exotic object, baking in too many assumptions is bound to fail in edge cases.

The real test-case would be `return x instanceof HTMLElement` of course, but I see how that is difficult to test in an environment without `HTMLElement`, like nodejs.

Sanctuary-def's test of `HTMLElement` is good enough so that Sanctuary works client-side without requiring users to create a custom `HTMLElement` definition.
